File: 0001-two-sum/0001-two-sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {

        // cannot use same element twice

        unordered_map<int, int> num2idx;

        for (int i = 0; i < nums.size(); i++) {
            num2idx[nums[i]] = i;
        }

        for (int i = 0; i < nums.size(); i++) {
            int rest = target - nums[i];
            if (num2idx.find(rest) != num2idx.end() && num2idx[rest] != i)
                return {i, num2idx[rest]};
        }

        return {};
        
    }
};

File: 0001-two-sum/0001-two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        aMap = {}
        n = len(nums)
        
        for i in range(n):
            aMap[nums[i]] = i
        
        for i in range(n):
            rest = target - nums[i]
            if rest in aMap and aMap[rest] != i:
                return [aMap[rest], i]
            
        return []